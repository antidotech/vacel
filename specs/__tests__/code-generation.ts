import { wrap as wrapRaw } from 'jest-snapshot-serializer-raw'
import { testdocs } from '../../__tests__/helpers/snapshot'
import { generate } from '../../src/lib'
import { Tokenizer } from '../../src/parser/tokenizer'

for (const testdoc of testdocs) {
  for (const testcase of testdoc.basic) {
    describe(`(${testdoc.name})`, () => {
      const ast = testdoc.parse.call(testdoc.parse, testcase.code)
      const { code } = generate(ast)

      it(testcase.name, () => {
        if (
          testdoc.name === 'BinaryExpression' &&
          testcase.name === 'multiple'
        ) {
          // no test because of parens auto insertion
          return
        }

        const output = new Tokenizer(code)
          .tokenize()
          .filter((token) => token.type !== 'comment')
          .map((token) => token.value)

        const input = new Tokenizer(testcase.code)
          .tokenize()
          .filter((token) => token.type !== 'comment')
          .map((token) => token.value)

        expect(output).toStrictEqual(input)
      })

      const formatted = {
        long: generate(ast, { printWidth: Infinity }).code,
        short: generate(ast, { printWidth: 0 }).code,
      }

      if (formatted.short === formatted.long) {
        it('format', () => {
          expect(wrap(testcase.code, formatted.long)).toMatchSnapshot()
        })
      } else {
        it('format<long>', () => {
          expect(wrap(testcase.code, formatted.long)).toMatchSnapshot()
        })

        it('format<short>', () => {
          expect(wrap(testcase.code, formatted.short)).toMatchSnapshot()
        })
      }
    })
  }
}

const wrap = (input: string, output: string) =>
  wrapRaw(
    `
---------- (input) ---------------------------------------------------------------
${input}

---------- (output) --------------------------------------------------------------
${output}

`.trim()
  )
