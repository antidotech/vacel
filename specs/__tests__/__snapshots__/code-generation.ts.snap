// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(AclStatement) format 1`] = `
---------- (input) ---------------------------------------------------------------
acl aaa {
  "127.0.0.1";
  "localhost";
}

---------- (output) --------------------------------------------------------------
acl aaa {
  "127.0.0.1";
  "localhost";
}
`;

exports[`(AddStatement) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
add var.vacel = "vacel";

---------- (output) --------------------------------------------------------------
add var.vacel = "vacel";
`;

exports[`(AddStatement) format<long> 2`] = `
---------- (input) ---------------------------------------------------------------
add var.vacel += "vacel";

---------- (output) --------------------------------------------------------------
add var.vacel += "vacel";
`;

exports[`(AddStatement) format<long> 3`] = `
---------- (input) ---------------------------------------------------------------
add var.vacel -= "vacel";

---------- (output) --------------------------------------------------------------
add var.vacel -= "vacel";
`;

exports[`(AddStatement) format<long> 4`] = `
---------- (input) ---------------------------------------------------------------
add var.vacel *= "vacel";

---------- (output) --------------------------------------------------------------
add var.vacel *= "vacel";
`;

exports[`(AddStatement) format<long> 5`] = `
---------- (input) ---------------------------------------------------------------
add var.vacel /= "vacel";

---------- (output) --------------------------------------------------------------
add var.vacel /= "vacel";
`;

exports[`(AddStatement) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
add var.vacel = "vacel";

---------- (output) --------------------------------------------------------------
add var.vacel =
  "vacel";
`;

exports[`(AddStatement) format<short> 2`] = `
---------- (input) ---------------------------------------------------------------
add var.vacel += "vacel";

---------- (output) --------------------------------------------------------------
add var.vacel +=
  "vacel";
`;

exports[`(AddStatement) format<short> 3`] = `
---------- (input) ---------------------------------------------------------------
add var.vacel -= "vacel";

---------- (output) --------------------------------------------------------------
add var.vacel -=
  "vacel";
`;

exports[`(AddStatement) format<short> 4`] = `
---------- (input) ---------------------------------------------------------------
add var.vacel *= "vacel";

---------- (output) --------------------------------------------------------------
add var.vacel *=
  "vacel";
`;

exports[`(AddStatement) format<short> 5`] = `
---------- (input) ---------------------------------------------------------------
add var.vacel /= "vacel";

---------- (output) --------------------------------------------------------------
add var.vacel /=
  "vacel";
`;

exports[`(BackendStatement) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
backend vaceline_backend {
  .connect_timeout = 1s;
  .dynamic = true;
  .port = "443";
  .host = "vaceline.io";
  .first_byte_timeout = 5s;
  .max_connections = 100;
  .between_bytes_timeout = 10s;
  .ssl = true;
  .probe = {
    .request = "GET /healthcheck HTTP/1.1" "Host: vaceline.io" "Connection: close" "User-Agent: Vaceline healthcheck";
    .threshold = 1;
    .window = 2;
    .timeout = 10s;
    .initial = 1;
    .expected_response = 200;
    .interval = 60s;
  }
}

---------- (output) --------------------------------------------------------------
backend vaceline_backend {
  .connect_timeout = 1s;
  .dynamic = true;
  .port = "443";
  .host = "vaceline.io";
  .first_byte_timeout = 5s;
  .max_connections = 100;
  .between_bytes_timeout = 10s;
  .ssl = true;
  .probe = {
    .request = "GET /healthcheck HTTP/1.1" "Host: vaceline.io" "Connection: close" "User-Agent: Vaceline healthcheck";
    .threshold = 1;
    .window = 2;
    .timeout = 10s;
    .initial = 1;
    .expected_response = 200;
    .interval = 60s;
  }
}
`;

exports[`(BackendStatement) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
backend vaceline_backend {
  .connect_timeout = 1s;
  .dynamic = true;
  .port = "443";
  .host = "vaceline.io";
  .first_byte_timeout = 5s;
  .max_connections = 100;
  .between_bytes_timeout = 10s;
  .ssl = true;
  .probe = {
    .request = "GET /healthcheck HTTP/1.1" "Host: vaceline.io" "Connection: close" "User-Agent: Vaceline healthcheck";
    .threshold = 1;
    .window = 2;
    .timeout = 10s;
    .initial = 1;
    .expected_response = 200;
    .interval = 60s;
  }
}

---------- (output) --------------------------------------------------------------
backend vaceline_backend {
  .connect_timeout = 1s;
  .dynamic = true;
  .port = "443";
  .host = "vaceline.io";
  .first_byte_timeout = 5s;
  .max_connections = 100;
  .between_bytes_timeout = 10s;
  .ssl = true;
  .probe = {
    .request = "GET /healthcheck HTTP/1.1"
      "Host: vaceline.io"
      "Connection: close"
      "User-Agent: Vaceline healthcheck";
    .threshold = 1;
    .window = 2;
    .timeout = 10s;
    .initial = 1;
    .expected_response = 200;
    .interval = 60s;
  }
}
`;

exports[`(BinaryExpression) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug == "1"

---------- (output) --------------------------------------------------------------
req.http.Debug == "1"
`;

exports[`(BinaryExpression) format<long> 2`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug != "1"

---------- (output) --------------------------------------------------------------
req.http.Debug != "1"
`;

exports[`(BinaryExpression) format<long> 3`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug ~ "1"

---------- (output) --------------------------------------------------------------
req.http.Debug ~ "1"
`;

exports[`(BinaryExpression) format<long> 4`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug == "1" == "2" == "3"

---------- (output) --------------------------------------------------------------
((req.http.Debug == "1") == "2") == "3"
`;

exports[`(BinaryExpression) format<long> 5`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug == "1" && "2" == "3"

---------- (output) --------------------------------------------------------------
req.http.Debug == "1" && "2" == "3"
`;

exports[`(BinaryExpression) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug == "1"

---------- (output) --------------------------------------------------------------
req
  .http
  .Debug ==
  "1"
`;

exports[`(BinaryExpression) format<short> 2`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug != "1"

---------- (output) --------------------------------------------------------------
req
  .http
  .Debug !=
  "1"
`;

exports[`(BinaryExpression) format<short> 3`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug ~ "1"

---------- (output) --------------------------------------------------------------
req
  .http
  .Debug ~
  "1"
`;

exports[`(BinaryExpression) format<short> 4`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug == "1" == "2" == "3"

---------- (output) --------------------------------------------------------------
((req
  .http
  .Debug ==
  "1") ==
  "2") ==
  "3"
`;

exports[`(BinaryExpression) format<short> 5`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug == "1" && "2" == "3"

---------- (output) --------------------------------------------------------------
req
  .http
  .Debug ==
  "1" &&
  "2" ==
    "3"
`;

exports[`(BooleanExpression) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
(req.http.Debug == "1")

---------- (output) --------------------------------------------------------------
(req.http.Debug == "1")
`;

exports[`(BooleanExpression) format<long> 2`] = `
---------- (input) ---------------------------------------------------------------
(var.Debug && "1")

---------- (output) --------------------------------------------------------------
(var.Debug && "1")
`;

exports[`(BooleanExpression) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
(req.http.Debug == "1")

---------- (output) --------------------------------------------------------------
(
  req
    .http
    .Debug ==
    "1"
)
`;

exports[`(BooleanExpression) format<short> 2`] = `
---------- (input) ---------------------------------------------------------------
(var.Debug && "1")

---------- (output) --------------------------------------------------------------
(
  var.Debug &&
    "1"
)
`;

exports[`(BooleanLiteral) format 1`] = `
---------- (input) ---------------------------------------------------------------
true

---------- (output) --------------------------------------------------------------
true
`;

exports[`(BooleanLiteral) format 2`] = `
---------- (input) ---------------------------------------------------------------
false

---------- (output) --------------------------------------------------------------
false
`;

exports[`(CallStatement) format 1`] = `
---------- (input) ---------------------------------------------------------------
call vaceline_func;

---------- (output) --------------------------------------------------------------
call vaceline_func;
`;

exports[`(ConcatExpression) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
"a" "b" "c"

---------- (output) --------------------------------------------------------------
"a" "b" "c"
`;

exports[`(ConcatExpression) format<long> 2`] = `
---------- (input) ---------------------------------------------------------------
"a" 1 req.http.A

---------- (output) --------------------------------------------------------------
"a" 1 req.http.A
`;

exports[`(ConcatExpression) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
"a" "b" "c"

---------- (output) --------------------------------------------------------------
"a"
  "b"
  "c"
`;

exports[`(ConcatExpression) format<short> 2`] = `
---------- (input) ---------------------------------------------------------------
"a" 1 req.http.A

---------- (output) --------------------------------------------------------------
"a"
  1
  req
    .http
    .A
`;

exports[`(DeclareStatement) format 1`] = `
---------- (input) ---------------------------------------------------------------
declare local var.Vaceline STRING;

---------- (output) --------------------------------------------------------------
declare local var.Vaceline STRING;
`;

exports[`(DeclareStatement) format 2`] = `
---------- (input) ---------------------------------------------------------------
declare local var.Vaceline BOOL;

---------- (output) --------------------------------------------------------------
declare local var.Vaceline BOOL;
`;

exports[`(DeclareStatement) format 3`] = `
---------- (input) ---------------------------------------------------------------
declare local var.Vaceline INTEGER;

---------- (output) --------------------------------------------------------------
declare local var.Vaceline INTEGER;
`;

exports[`(DeclareStatement) format 4`] = `
---------- (input) ---------------------------------------------------------------
declare local var.Vaceline FLOAT;

---------- (output) --------------------------------------------------------------
declare local var.Vaceline FLOAT;
`;

exports[`(DurationLiteral) format 1`] = `
---------- (input) ---------------------------------------------------------------
1ms

---------- (output) --------------------------------------------------------------
1ms
`;

exports[`(DurationLiteral) format 2`] = `
---------- (input) ---------------------------------------------------------------
1s

---------- (output) --------------------------------------------------------------
1s
`;

exports[`(DurationLiteral) format 3`] = `
---------- (input) ---------------------------------------------------------------
1m

---------- (output) --------------------------------------------------------------
1m
`;

exports[`(DurationLiteral) format 4`] = `
---------- (input) ---------------------------------------------------------------
1d

---------- (output) --------------------------------------------------------------
1d
`;

exports[`(DurationLiteral) format 5`] = `
---------- (input) ---------------------------------------------------------------
1y

---------- (output) --------------------------------------------------------------
1y
`;

exports[`(DurationLiteral) format 6`] = `
---------- (input) ---------------------------------------------------------------
1   m

---------- (output) --------------------------------------------------------------
1m
`;

exports[`(ErrorStatement) format 1`] = `
---------- (input) ---------------------------------------------------------------
error 404;

---------- (output) --------------------------------------------------------------
error 404;
`;

exports[`(ErrorStatement) format 2`] = `
---------- (input) ---------------------------------------------------------------
error 404 "Not Found";

---------- (output) --------------------------------------------------------------
error 404 "Not Found";
`;

exports[`(ErrorStatement) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
error 800 req.http.Debug;

---------- (output) --------------------------------------------------------------
error 800 req.http.Debug;
`;

exports[`(ErrorStatement) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
error 800 req.http.Debug;

---------- (output) --------------------------------------------------------------
error 800 req
  .http
  .Debug;
`;

exports[`(ExpressionStatement) format 1`] = `
---------- (input) ---------------------------------------------------------------
vaceline;

---------- (output) --------------------------------------------------------------
vaceline;
`;

exports[`(ExpressionStatement) format 2`] = `
---------- (input) ---------------------------------------------------------------
"string";

---------- (output) --------------------------------------------------------------
"string";
`;

exports[`(ExpressionStatement) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
req.http.Vaceline;

---------- (output) --------------------------------------------------------------
req.http.Vaceline;
`;

exports[`(ExpressionStatement) format<long> 2`] = `
---------- (input) ---------------------------------------------------------------
func(a);

---------- (output) --------------------------------------------------------------
func(a);
`;

exports[`(ExpressionStatement) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
req.http.Vaceline;

---------- (output) --------------------------------------------------------------
req
  .http
  .Vaceline;
`;

exports[`(ExpressionStatement) format<short> 2`] = `
---------- (input) ---------------------------------------------------------------
func(a);

---------- (output) --------------------------------------------------------------
func(
  a,
);
`;

exports[`(FunCallExpression) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
func("a")

---------- (output) --------------------------------------------------------------
func("a")
`;

exports[`(FunCallExpression) format<long> 2`] = `
---------- (input) ---------------------------------------------------------------
func  ("a")

---------- (output) --------------------------------------------------------------
func("a")
`;

exports[`(FunCallExpression) format<long> 3`] = `
---------- (input) ---------------------------------------------------------------
func("a", "b")

---------- (output) --------------------------------------------------------------
func("a", "b")
`;

exports[`(FunCallExpression) format<long> 4`] = `
---------- (input) ---------------------------------------------------------------
if("a", "b", "c")

---------- (output) --------------------------------------------------------------
if("a", "b", "c")
`;

exports[`(FunCallExpression) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
func("a")

---------- (output) --------------------------------------------------------------
func(
  "a",
)
`;

exports[`(FunCallExpression) format<short> 2`] = `
---------- (input) ---------------------------------------------------------------
func  ("a")

---------- (output) --------------------------------------------------------------
func(
  "a",
)
`;

exports[`(FunCallExpression) format<short> 3`] = `
---------- (input) ---------------------------------------------------------------
func("a", "b")

---------- (output) --------------------------------------------------------------
func(
  "a",
  "b",
)
`;

exports[`(FunCallExpression) format<short> 4`] = `
---------- (input) ---------------------------------------------------------------
if("a", "b", "c")

---------- (output) --------------------------------------------------------------
if(
  "a",
  "b",
  "c",
)
`;

exports[`(Identifier) format 1`] = `
---------- (input) ---------------------------------------------------------------
vaceline

---------- (output) --------------------------------------------------------------
vaceline
`;

exports[`(IfStatement) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
if (var.condition) {
  set var.value = 1;
}

---------- (output) --------------------------------------------------------------
if (var.condition) {
  set var.value = 1;
}
`;

exports[`(IfStatement) format<long> 2`] = `
---------- (input) ---------------------------------------------------------------
if (var.condition) {
  set var.value = 1;
} else {
  set var.value = 2;
}

---------- (output) --------------------------------------------------------------
if (var.condition) {
  set var.value = 1;
} else {
  set var.value = 2;
}
`;

exports[`(IfStatement) format<long> 3`] = `
---------- (input) ---------------------------------------------------------------
if (var.condition) {
  set var.value = 1;
} else if (!var.condition) {
  set var.value = 2;
} else {
  set var.value = 0;
}

---------- (output) --------------------------------------------------------------
if (var.condition) {
  set var.value = 1;
} else if (!var.condition) {
  set var.value = 2;
} else {
  set var.value = 0;
}
`;

exports[`(IfStatement) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
if (var.condition) {
  set var.value = 1;
}

---------- (output) --------------------------------------------------------------
if (
  var.condition
) {
  set var.value =
    1;
}
`;

exports[`(IfStatement) format<short> 2`] = `
---------- (input) ---------------------------------------------------------------
if (var.condition) {
  set var.value = 1;
} else {
  set var.value = 2;
}

---------- (output) --------------------------------------------------------------
if (
  var.condition
) {
  set var.value =
    1;
} else {
  set var.value =
    2;
}
`;

exports[`(IfStatement) format<short> 3`] = `
---------- (input) ---------------------------------------------------------------
if (var.condition) {
  set var.value = 1;
} else if (!var.condition) {
  set var.value = 2;
} else {
  set var.value = 0;
}

---------- (output) --------------------------------------------------------------
if (
  var.condition
) {
  set var.value =
    1;
} else if (
  !var.condition
) {
  set var.value =
    2;
} else {
  set var.value =
    0;
}
`;

exports[`(ImportStatement) format 1`] = `
---------- (input) ---------------------------------------------------------------
import module;

---------- (output) --------------------------------------------------------------
import module;
`;

exports[`(IncludeStatement) format 1`] = `
---------- (input) ---------------------------------------------------------------
include "external_library";

---------- (output) --------------------------------------------------------------
include "external_library";
`;

exports[`(Ip) format 1`] = `
---------- (input) ---------------------------------------------------------------
"localhost"

---------- (output) --------------------------------------------------------------
"localhost"
`;

exports[`(Ip) format 2`] = `
---------- (input) ---------------------------------------------------------------
"192.0.2.0"

---------- (output) --------------------------------------------------------------
"192.0.2.0"
`;

exports[`(Ip) format 3`] = `
---------- (input) ---------------------------------------------------------------
"192.0.2.0"/16

---------- (output) --------------------------------------------------------------
"192.0.2.0"/16
`;

exports[`(Ip) format 4`] = `
---------- (input) ---------------------------------------------------------------
"2001:db8::1"

---------- (output) --------------------------------------------------------------
"2001:db8::1"
`;

exports[`(Ip) format 5`] = `
---------- (input) ---------------------------------------------------------------
"2001:db8::1"/16

---------- (output) --------------------------------------------------------------
"2001:db8::1"/16
`;

exports[`(Ip) format 6`] = `
---------- (input) ---------------------------------------------------------------
"2002:c000:0204::"

---------- (output) --------------------------------------------------------------
"2002:c000:0204::"
`;

exports[`(Ip) format 7`] = `
---------- (input) ---------------------------------------------------------------
"::FFFF:192.0.2.4"

---------- (output) --------------------------------------------------------------
"::FFFF:192.0.2.4"
`;

exports[`(Ip) format 8`] = `
---------- (input) ---------------------------------------------------------------
"::1"

---------- (output) --------------------------------------------------------------
"::1"
`;

exports[`(Ip) format 9`] = `
---------- (input) ---------------------------------------------------------------
"::"

---------- (output) --------------------------------------------------------------
"::"
`;

exports[`(LogStatement) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
log "log" req.service_id "vaceline-log :: "
  {"	timestamp_us:"} time.start.usec
  {"	host:"} regsuball(req.http.X-Forwarded-Host, {"	"}, "");

---------- (output) --------------------------------------------------------------
log "log" req.service_id "vaceline-log :: " {"	timestamp_us:"} time.start.usec {"	host:"} regsuball(req.http.X-Forwarded-Host, {"	"}, "");
`;

exports[`(LogStatement) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
log "log" req.service_id "vaceline-log :: "
  {"	timestamp_us:"} time.start.usec
  {"	host:"} regsuball(req.http.X-Forwarded-Host, {"	"}, "");

---------- (output) --------------------------------------------------------------
log "log"
  req.service_id
  "vaceline-log :: "
  {"	timestamp_us:"}
  time
    .start
    .usec
  {"	host:"}
  regsuball(
    req
      .http
      .X-Forwarded-Host,
    {"	"},
    "",
  );
`;

exports[`(LogicalExpression) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug && "1"

---------- (output) --------------------------------------------------------------
req.http.Debug && "1"
`;

exports[`(LogicalExpression) format<long> 2`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug || "1"

---------- (output) --------------------------------------------------------------
req.http.Debug || "1"
`;

exports[`(LogicalExpression) format<long> 3`] = `
---------- (input) ---------------------------------------------------------------
debug && "1" || "2"

---------- (output) --------------------------------------------------------------
(debug && "1") || "2"
`;

exports[`(LogicalExpression) format<long> 4`] = `
---------- (input) ---------------------------------------------------------------
debug || "1" && "2"

---------- (output) --------------------------------------------------------------
debug || ("1" && "2")
`;

exports[`(LogicalExpression) format<long> 5`] = `
---------- (input) ---------------------------------------------------------------
debug == "1" && "2"

---------- (output) --------------------------------------------------------------
debug == "1" && "2"
`;

exports[`(LogicalExpression) format<long> 6`] = `
---------- (input) ---------------------------------------------------------------
debug == "1" || "2"

---------- (output) --------------------------------------------------------------
debug == "1" || "2"
`;

exports[`(LogicalExpression) format<long> 7`] = `
---------- (input) ---------------------------------------------------------------
debug && "1" == "2"

---------- (output) --------------------------------------------------------------
debug && "1" == "2"
`;

exports[`(LogicalExpression) format<long> 8`] = `
---------- (input) ---------------------------------------------------------------
debug || "1" == "2"

---------- (output) --------------------------------------------------------------
debug || "1" == "2"
`;

exports[`(LogicalExpression) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug && "1"

---------- (output) --------------------------------------------------------------
req
  .http
  .Debug &&
  "1"
`;

exports[`(LogicalExpression) format<short> 2`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug || "1"

---------- (output) --------------------------------------------------------------
req
  .http
  .Debug ||
  "1"
`;

exports[`(LogicalExpression) format<short> 3`] = `
---------- (input) ---------------------------------------------------------------
debug && "1" || "2"

---------- (output) --------------------------------------------------------------
(debug &&
  "1") ||
  "2"
`;

exports[`(LogicalExpression) format<short> 4`] = `
---------- (input) ---------------------------------------------------------------
debug || "1" && "2"

---------- (output) --------------------------------------------------------------
debug ||
  ("1" &&
    "2")
`;

exports[`(LogicalExpression) format<short> 5`] = `
---------- (input) ---------------------------------------------------------------
debug == "1" && "2"

---------- (output) --------------------------------------------------------------
debug ==
  "1" &&
  "2"
`;

exports[`(LogicalExpression) format<short> 6`] = `
---------- (input) ---------------------------------------------------------------
debug == "1" || "2"

---------- (output) --------------------------------------------------------------
debug ==
  "1" ||
  "2"
`;

exports[`(LogicalExpression) format<short> 7`] = `
---------- (input) ---------------------------------------------------------------
debug && "1" == "2"

---------- (output) --------------------------------------------------------------
debug &&
  "1" ==
    "2"
`;

exports[`(LogicalExpression) format<short> 8`] = `
---------- (input) ---------------------------------------------------------------
debug || "1" == "2"

---------- (output) --------------------------------------------------------------
debug ||
  "1" ==
    "2"
`;

exports[`(Member) format 1`] = `
---------- (input) ---------------------------------------------------------------
req.backend

---------- (output) --------------------------------------------------------------
req.backend
`;

exports[`(Member) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug

---------- (output) --------------------------------------------------------------
req.http.Debug
`;

exports[`(Member) format<long> 2`] = `
---------- (input) ---------------------------------------------------------------
req.backend == backend

---------- (output) --------------------------------------------------------------
req.backend == backend
`;

exports[`(Member) format<long> 3`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug == "debug"

---------- (output) --------------------------------------------------------------
req.http.Debug == "debug"
`;

exports[`(Member) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug

---------- (output) --------------------------------------------------------------
req
  .http
  .Debug
`;

exports[`(Member) format<short> 2`] = `
---------- (input) ---------------------------------------------------------------
req.backend == backend

---------- (output) --------------------------------------------------------------
req.backend ==
  backend
`;

exports[`(Member) format<short> 3`] = `
---------- (input) ---------------------------------------------------------------
req.http.Debug == "debug"

---------- (output) --------------------------------------------------------------
req
  .http
  .Debug ==
  "debug"
`;

exports[`(MultilineLiteral) format 1`] = `
---------- (input) ---------------------------------------------------------------
{"
  vaceline
"}

---------- (output) --------------------------------------------------------------
{"
  vaceline
"}
`;

exports[`(NumericLiteral) format 1`] = `
---------- (input) ---------------------------------------------------------------
10

---------- (output) --------------------------------------------------------------
10
`;

exports[`(NumericLiteral) format 2`] = `
---------- (input) ---------------------------------------------------------------
0

---------- (output) --------------------------------------------------------------
0
`;

exports[`(Program) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
sub vcl_recv {
  declare local var.vacelinDebug STRING;

  if (req.http.Vacelin:Some == "Some") {
    add var.vacelinDebug = "Vacelin Debug Log";
  }
}

---------- (output) --------------------------------------------------------------
sub vcl_recv {
  declare local var.vacelinDebug STRING;
  if (req.http.Vacelin:Some == "Some") {
    add var.vacelinDebug = "Vacelin Debug Log";
  }
}
`;

exports[`(Program) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
sub vcl_recv {
  declare local var.vacelinDebug STRING;

  if (req.http.Vacelin:Some == "Some") {
    add var.vacelinDebug = "Vacelin Debug Log";
  }
}

---------- (output) --------------------------------------------------------------
sub vcl_recv {
  declare local var.vacelinDebug STRING;
  if (
    req
      .http
      .Vacelin:Some ==
      "Some"
  ) {
    add var.vacelinDebug =
      "Vacelin Debug Log";
  }
}
`;

exports[`(RestartStatement) format 1`] = `
---------- (input) ---------------------------------------------------------------
restart;

---------- (output) --------------------------------------------------------------
restart;
`;

exports[`(ReturnStatement) format 1`] = `
---------- (input) ---------------------------------------------------------------
return (pass);

---------- (output) --------------------------------------------------------------
return (pass);
`;

exports[`(SetStatement) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
set var.vacel = "vacel";

---------- (output) --------------------------------------------------------------
set var.vacel = "vacel";
`;

exports[`(SetStatement) format<long> 2`] = `
---------- (input) ---------------------------------------------------------------
set var.vacel += "vacel";

---------- (output) --------------------------------------------------------------
set var.vacel += "vacel";
`;

exports[`(SetStatement) format<long> 3`] = `
---------- (input) ---------------------------------------------------------------
set var.vacel -= "vacel";

---------- (output) --------------------------------------------------------------
set var.vacel -= "vacel";
`;

exports[`(SetStatement) format<long> 4`] = `
---------- (input) ---------------------------------------------------------------
set var.vacel *= "vacel";

---------- (output) --------------------------------------------------------------
set var.vacel *= "vacel";
`;

exports[`(SetStatement) format<long> 5`] = `
---------- (input) ---------------------------------------------------------------
set var.vacel *= "vacel";

---------- (output) --------------------------------------------------------------
set var.vacel *= "vacel";
`;

exports[`(SetStatement) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
set var.vacel = "vacel";

---------- (output) --------------------------------------------------------------
set var.vacel =
  "vacel";
`;

exports[`(SetStatement) format<short> 2`] = `
---------- (input) ---------------------------------------------------------------
set var.vacel += "vacel";

---------- (output) --------------------------------------------------------------
set var.vacel +=
  "vacel";
`;

exports[`(SetStatement) format<short> 3`] = `
---------- (input) ---------------------------------------------------------------
set var.vacel -= "vacel";

---------- (output) --------------------------------------------------------------
set var.vacel -=
  "vacel";
`;

exports[`(SetStatement) format<short> 4`] = `
---------- (input) ---------------------------------------------------------------
set var.vacel *= "vacel";

---------- (output) --------------------------------------------------------------
set var.vacel *=
  "vacel";
`;

exports[`(SetStatement) format<short> 5`] = `
---------- (input) ---------------------------------------------------------------
set var.vacel *= "vacel";

---------- (output) --------------------------------------------------------------
set var.vacel *=
  "vacel";
`;

exports[`(StringLiteral) format 1`] = `
---------- (input) ---------------------------------------------------------------
"string"

---------- (output) --------------------------------------------------------------
"string"
`;

exports[`(StringLiteral) format 2`] = `
---------- (input) ---------------------------------------------------------------
"# not a comment"

---------- (output) --------------------------------------------------------------
"# not a comment"
`;

exports[`(StringLiteral) format 3`] = `
---------- (input) ---------------------------------------------------------------
"/* not an inline comment */"

---------- (output) --------------------------------------------------------------
"/* not an inline comment */"
`;

exports[`(SubroutineStatement) format<long> 1`] = `
---------- (input) ---------------------------------------------------------------
sub vcl_recv {
  set vaceline = "aaa";
}

---------- (output) --------------------------------------------------------------
sub vcl_recv {
  set vaceline = "aaa";
}
`;

exports[`(SubroutineStatement) format<short> 1`] = `
---------- (input) ---------------------------------------------------------------
sub vcl_recv {
  set vaceline = "aaa";
}

---------- (output) --------------------------------------------------------------
sub vcl_recv {
  set vaceline =
    "aaa";
}
`;

exports[`(SyntheticStatement) format 1`] = `
---------- (input) ---------------------------------------------------------------
synthetic "";

---------- (output) --------------------------------------------------------------
synthetic "";
`;

exports[`(TableStatement) format 1`] = `
---------- (input) ---------------------------------------------------------------
table vaceline_table {
  "key": "value"
}

---------- (output) --------------------------------------------------------------
table vaceline_table {
  "key":"value"
}
`;

exports[`(TableStatement) format 2`] = `
---------- (input) ---------------------------------------------------------------
table vaceline_table {
  "key": "value",
  "key2": "value2"
}

---------- (output) --------------------------------------------------------------
table vaceline_table {
  "key":"value",
  "key2":"value2"
}
`;

exports[`(UnaryExpression) format 1`] = `
---------- (input) ---------------------------------------------------------------
!true

---------- (output) --------------------------------------------------------------
!true
`;

exports[`(UnaryExpression) format 2`] = `
---------- (input) ---------------------------------------------------------------
!condition

---------- (output) --------------------------------------------------------------
!condition
`;

exports[`(UnsetStatement) format 1`] = `
---------- (input) ---------------------------------------------------------------
unset value;

---------- (output) --------------------------------------------------------------
unset value;
`;

exports[`(UnsetStatement) format 2`] = `
---------- (input) ---------------------------------------------------------------
unset var.value;

---------- (output) --------------------------------------------------------------
unset var.value;
`;

exports[`(ValuePair) format 1`] = `
---------- (input) ---------------------------------------------------------------
Cookie:Vacelin

---------- (output) --------------------------------------------------------------
Cookie:Vacelin
`;
